Added improved example for ConvLSTM-based next-frame prediction

amogh7joshi opened the pull request:
This example aims to follow the *Improve next-frame prediction with ConvLSTM* call for contributions by creating a new example with the `Moving MNIST` dataset and improved explanations.

fchollet commented:
Thanks for the PR! Do you have a live Colab notebook so I can check out the visualizations?

amogh7joshi commented:
Yup, you can check it out [right here](https://colab.research.google.com/drive/1V_rzxWsxa8Y2teAt6LQYvClntLSZ-VEw?usp=sharing). BTW, at the moment the notebook is just running based on 5 epochs of training (since this is supposed to be a lightweight example) so I haven't done any rigorous training. Let me know if we need some better outputs, I can look into either increased training, increasing the dataset size, or adjusting the model/loss.

fchollet commented:
It seems that you need to train the model much longer, and also tune the hyperparameters (note: you may not need BatchNorm). It's fine if the training takes a long time, we can just save the output picture and show it in the notebook, while restricting the number of epochs in the published notebook.

I recommend making this example replace the old one we had (rather than adding a new one).

amogh7joshi commented:
Cool, I've updated the notebook with new results. The way I had it previously was that each predicted frame would be appended to the original sequence and new predictions made off of that, which I don't think was exactly what the paper did, so now it's just that I take the original sequence and make 10 predictions off of extended versions of _that_, which is what is mentioned for the Moving MNIST dataset in the paper.

I switched out the hyperparams and also switched to only one BatchNorm layer, and w/ 20 epochs, I reach comparable visual results to the original paper. Let me know if it's good, and I'll update the original `.py` file as such.

fchollet commented:
Thanks! The generate figures aren't very clear; we should make it clear what is being generated, and from what. Perhaps a gif animation would also be helpful.

amogh7joshi commented:
Ok, I've removed the original example and replaced it with the new one (I'll remove the notebook/md file at the end). The original one was called `conv_lstm.py`, but I've called the new one `next_frame_prediction.py` considering that's more apt for the actual goal of the example, although if you want to keep the previous name, let me know.

On the topic of descriptions, what would be the best way to go about this? I could just pick a few examples and then instead of a figure displaying each individual frame, I could just create a few GIFs showing the progression of the sequence of frames in each, like you mentioned, which I think could be a decent visualization. Alternatively, I could retain what's already there, and then just add the GIFs after it, which would allow both a visualization of the actual frame-by-frame predictions, as well as a final product. Personally, I'm leaning towards the second one, so I'll start hacking around making those GIFs, and let me know if you have a different idea.

fchollet commented:
> although if you want to keep the previous name, let me know

Let's keep the previous filename, because we need to keep the URL location.

fchollet commented:
> Personally, I'm leaning towards the second one, so I'll start hacking around making those GIFs, and let me know if you have a different idea.

Sounds fine. One things will be very important is providing detailed labels for the figures, so the reader immediately knows what they're looking at (I didn't).

fchollet commented on examples/vision/next_frame_prediction.py:
```diff
@@ -0,0 +1,233 @@
+
+import numpy as np
+import matplotlib.pyplot as plt
+from sklearn.model_selection import train_test_split
```
Do this by hand rather than depending on sklearn for it

fchollet commented on examples/vision/next_frame_prediction.py:
```diff
@@ -0,0 +1,233 @@
+# We'll define a helper function to shift the frames, where
+# `X` is frames 0 to n - 1, and `y` is frames 1 to n.
+def create_shifted_frames(data):
+    X = data[:, 0 : data.shape[1] - 1, :, :]
```
Always user lowercase for variable names

fchollet commented on examples/vision/next_frame_prediction.py:
```diff
@@ -0,0 +1,233 @@
+"""
+
+# Download and load the dataset.
+fpath = tf.keras.utils.get_file(
```
You already import keras so no need for the `tf.` here

amogh7joshi commented:
I've updated the Colab notebook as well as the actual tutorial `.py` file with GIF visualizations at the end and better labeling, as well as the different reviews that were suggested. Let me know if I should add any more descriptions to the output visualizations.

fchollet reviewed (commented):
I've added some copyedits. Please pull them and then add the generate files -- everything is looking great. Thank you!

amogh7joshi commented:
Great! Quick question–does the `autogen.py` file actually _run_ the Jupyter Notebook after creating it? Since I don't have a local GPU (I've been using Colab's), I can't exactly run the 20 or even a shortened 10 epoch training without it taking an infeasible amount of time. 

If it does, please let me know what I can do to work around this.

fchollet commented:
> Since I don't have a local GPU (I've been using Colab's), I can't exactly run the 20 or even a shortened 10 epoch training without it taking an infeasible amount of time.

I would recommend using a cloud VM (like a GCP VM) with a GPU to run the notebook.

Another thing you could do:

- Comment out the `fit()` line (which is what takes time).
- Run `autogen` on your laptop to get the md and ipynb.
- Take a look at what the md file contains. Take note the generated image files as well.
- Edit them by hand to fill in the output that you obtained in Colab (both the md file and the image files), basically the result you'd have obtained without commenting out the line
- Edit the ipynb (you just need tio uncomment the line you commented)
- push the resulting files

If you do this, be very careful that you're making the right edits.

amogh7joshi commented:
Alright, I've added the generated files. I opted for the second option and have proofread both the notebook and the Markdown file a few times, and then also generated the actual site for some final checks.

I think it should be good to go, but let me know if the final rendered images on the actual site rendering (the markdown file, I suppose) need any more improvements.

fchollet reviewed (approved):
Thank you!

I'll merge and check it out. If there are further changes needed let's do that in a new PR.

fchollet commented:
Everything is looking good. Thanks again for the great contribution! 👍

amogh7joshi commented:
Thanks a lot for your help! Looking forward to seeing the new example up on `keras.io`.
